amgcl/detail/qr.hpp

```diff
@@ -223,10 +223,10 @@ class QR {
             const int row_stride = (Order == row_major ? nmax : 1);
             const int col_stride = (Order == row_major ? 1 : m);
 
-            for(int i = 0, ii = 0; i < n; ++i, ii += row_stride + col_stride)
+            for(int i = 0, ii = 0; i < m; ++i, ii += row_stride + col_stride)
```

Comment:
Just thinking aloud here: we solve the system `QRx=f`, or `Qy=f`, `Rx=y`. Matrices `Q` and `R` have dimensions (rows x cols) `[m x k]` and `[k x n]` correspondingly. `Q` is orthogonal (`Q Q^T=I`), so we solve `Qy=f` by `y = Q^Tf`. That is the first loop here. Then we solve `Rx=y` using the fact that `R` is triangular. This is only the case for tall matrices though (`m > n`), and the function should work unchanged in this case.

For wide matrices, it seems the approach should be different, see [1], p.7-8. Factorization `QR=A^T` should be used instead of `QR=A`, and then we should solve `R^T Q^Tx = b`.

[1] CS322 Lecture Notes: QR factorization and orthogonal transformations, Steve Marschner Cornell University 2–5 March 2007. http://www.cs.cornell.edu/courses/cs322/2007sp/notes/qr.pdf